### ISSUE TITLE
Misleading error message about "call to unsafe function"

### ISSUE THREAD

**Issue by ruuda** [rust-lang/rust#44674]
The following code generates a misleading error message on Rust 1.20.0:

```rust
fn main() {
    use std::mem;
    unsafe {
        const MAGIC: i32 = mem::transmute([1u8, 2, 3, 4]);
        println!("{}", MAGIC);
    }
}
```

Observed output:

```
error[E0133]: call to unsafe function requires unsafe function or block
 --> src/main.rs:4:28
  |
4 |         const MAGIC: i32 = mem::transmute([1u8, 2, 3, 4]);
  |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ call to unsafe function
```

It is misleading, because the call is in fact inside an unsafe block. Expected output:

    error[E0015]: calls in constants are limited to struct and enum constructors

On the current nightly the problem is even worse because the compiler also warns about an unused unsafe block, while at the same time telling that an unsafe block is required.

**Comment by arielb1** [comment]
The error is correct - every item is a scope of its own. You'll have to put an unsafe block around the *function*:

```Rust
fn main() {
    use std::mem;
    
        const MAGIC: i32 = unsafe {
            mem::transmute([1u8, 2, 3, 4])
        };
        println!("{}", MAGIC);
}
```

**Comment by Mark-Simulacrum** [comment]
This is a duplicate of https://github.com/rust-lang/rust/issues/35716. Closing.
